### ISSUE TITLE
`BloomFilterSize` makes daemon forget about stuff on restarts when using ipfs/go-s3-ds

### ISSUE THREAD

**Issue by fiatjaf** [ipfs/go-ipfs#5235]
#### Version information:
go-ipfs version: 0.4.16-rc1-7e8f6c9
Repo version: 7
System version: amd64/linux
Golang version: go1.10
(built from latest master)

#### Type:
Bug

#### Description:
I was trying to get an S3-backed IPFS repo that I would use to just pin some selected files of mine, thus I didn't want it to participate in distributing any other content besides what I had pinned on it and I wouldn't use it to `get` any resource, it would just live there hosting my handful of pinned files.

I managed to get https://github.com/ipfs/go-ds-s3 working by modifying `repo/fsrepo/datastores.go` and everything worked fine during my tests.

Then, as I noticed (through log statements carefully inserted) the datastore plugin was doing a huge amount of `.Has()` calls to S3 (which I didn't want, as they would cost a lot of money for my humble purposes), I proceeded, guided by @whyrusleeping on IRC, to perform some modifications to `config` so my node would just answer "no" to all DHT queries from peers and "yes" to just the content it had pinned (as you see, I don't know exactly how peers communicate, but you get the idea).

The configuration change that made the `.Has()` calls stop was to set `BloomFilterSize` to a non-zero value. And it all worked fine.

Now comes the problem:

1. start the daemon
2. `echo 'kalimbaharkandant' | ipfs add` => QmTaWdr426NPe2wfzMfqLsbbA3CZEchBv2wWQBefNDSGdB
3. `ipfs cat QmTaWdr426NPe2wfzMfqLsbbA3CZEchBv2wWQBefNDSGdB` => kalimbaharkandant
4. restart the daemon
5. `ipfs cat QmTaWdr426NPe2wfzMfqLsbbA3CZEchBv2wWQBefNDSGdB` => (never completes)

These steps are all done on the same machine.

I've tried to set the `BloomFilterSize` using the default flatfs+leveldb datastore config but the problem doesn't occur there.

My `config` file as of now:

```
{
  "Identity": {
    "PeerID": "QmT2rxdjjRKTxfPFrmwY6MgDEgZsv83gGeRqQybu7YGTXP",
    "PrivKey": "<redacted>"
  },
  "Datastore": {
    "StorageMax": "10GB",
    "StorageGCWatermark": 90,
    "GCPeriod": "1h",
    "Spec": {
      "mounts": [
        {
          "accessKey": "<redacted>",
          "bucket": "fiatjaf-ipfs",
          "mountpoint": "/blocks",
          "secretKey": "<redacted>",
          "type": "s3ds"
        },
        {
          "child": {
            "compression": "none",
            "path": "datastore",
            "type": "levelds"
          },
          "mountpoint": "/",
          "prefix": "leveldb.datastore",
          "type": "measure"
        }
      ],
      "type": "mount"
    },
    "HashOnRead": false,
    "BloomFilterSize": 10000000
  },
  "Addresses": {
    "Swarm": [
      "/ip4/0.0.0.0/tcp/4001",
      "/ip6/::/tcp/4001"
    ],
    "Announce": [],
    "NoAnnounce": [],
    "API": "/ip4/127.0.0.1/tcp/5001",
    "Gateway": "/ip4/127.0.0.1/tcp/8080"
  },
  "Mounts": {
    "IPFS": "/ipfs",
    "IPNS": "/ipns",
    "FuseAllowOther": false
  },
  "Discovery": {
    "MDNS": {
      "Enabled": true,
      "Interval": 10
    }
  },
  "Routing": {
    "Type": "dht"
  },
  "Ipns": {
    "RepublishPeriod": "",
    "RecordLifetime": "",
    "ResolveCacheSize": 128
  },
  "Bootstrap": [
    "/dnsaddr/bootstrap.libp2p.io/ipfs/QmNnooDu7bfjPFoTZYxMNLWUQJyrVwtbZg5gBMjTezGAJN",
    "/dnsaddr/bootstrap.libp2p.io/ipfs/QmQCU2EcMqAqQPR2i9bChDtGNJchTbq5TbXJJ16u19uLTa",
    "/dnsaddr/bootstrap.libp2p.io/ipfs/QmbLHAnMoJPWSCR5Zhtx6BHJX9KiKNN6tpvbUcqanj75Nb",
    "/dnsaddr/bootstrap.libp2p.io/ipfs/QmcZf59bWwK5XFi76CZX8cbJ4BhTzzA3gU1ZjYZcYW3dwt",
    "/ip4/104.131.131.82/tcp/4001/ipfs/QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ",
    "/ip4/104.236.179.241/tcp/4001/ipfs/QmSoLPppuBtQSGwKDZT2M73ULpjvfd3aZ6ha4oFGL1KrGM",
    "/ip4/128.199.219.111/tcp/4001/ipfs/QmSoLSafTMBsPKadTEgaXctDQVcqN88CNLHXMkTNwMKPnu",
    "/ip4/104.236.76.40/tcp/4001/ipfs/QmSoLV4Bbm51jM9C4gDYZQ9Cy3U6aXMJDAbzgu2fzaDs64",
    "/ip4/178.62.158.247/tcp/4001/ipfs/QmSoLer265NRgSp2LA3dPaeykiS1J6DifTC88f5uVQKNAd",
    "/ip6/2604:a880:1:20::203:d001/tcp/4001/ipfs/QmSoLPppuBtQSGwKDZT2M73ULpjvfd3aZ6ha4oFGL1KrGM",
    "/ip6/2400:6180:0:d0::151:6001/tcp/4001/ipfs/QmSoLSafTMBsPKadTEgaXctDQVcqN88CNLHXMkTNwMKPnu",
    "/ip6/2604:a880:800:10::4a:5001/tcp/4001/ipfs/QmSoLV4Bbm51jM9C4gDYZQ9Cy3U6aXMJDAbzgu2fzaDs64",
    "/ip6/2a03:b0c0:0:1010::23:1001/tcp/4001/ipfs/QmSoLer265NRgSp2LA3dPaeykiS1J6DifTC88f5uVQKNAd"
  ],
  "Gateway": {
    "HTTPHeaders": {
      "Access-Control-Allow-Headers": [
        "X-Requested-With",
        "Range"
      ],
      "Access-Control-Allow-Methods": [
        "GET"
      ],
      "Access-Control-Allow-Origin": [
        "*"
      ]
    },
    "RootRedirect": "",
    "Writable": false,
    "PathPrefixes": []
  },
  "API": {
    "HTTPHeaders": {
      "Server": [
        "go-ipfs/0.4.16-rc1"
      ]
    }
  },
  "Swarm": {
    "AddrFilters": null,
    "DisableBandwidthMetrics": false,
    "DisableNatPortMap": false,
    "DisableRelay": false,
    "EnableRelayHop": false,
    "ConnMgr": {
      "Type": "basic",
      "LowWater": 600,
      "HighWater": 900,
      "GracePeriod": "20s"
    }
  },
  "Reprovider": {
    "Interval": "12h",
    "Strategy": "all"
  },
  "Experimental": {
    "FilestoreEnabled": false,
    "ShardingEnabled": false,
    "Libp2pStreamMounting": false
  }
}
```

I can share the code I modified to implement the datastore support, but I don't think it is relevant.

**Comment by whyrusleeping** [comment]
@Kubuxu or @magik6k any idea here?

**Comment by fiatjaf** [comment]
That can't be the problem. I was testing this with a datastore that had nothing on it. I basically just compiled go-ipfs with the s3 datastore support, initialized the repo and run the steps above.

Without the bloom filter, I get nonstop `Has()` calls (even without any files) going on forever. With the bloom filter I guess I get one `Query()` (not sure, probably one `Has()`?) and nothing more on restart -- however it doesn't work.

I also tried to reproduce using the default flatfs, but couldn't.

I don't know what is inside the leveldb on `/datastore` (as opposed to `/blocks`) so maybe the problem is in these two interfacing with each other? I've configured so I would still have it on my disk (while `/blocks` is on S3) because I thought it was an index or something like that, but what is it?
